packager/media/formats/mp2t/mpeg1_header.cc

```diff
@@ -0,0 +1,232 @@
+
+  DCHECK_LT(btr_idx, kMpeg1BitrateTableSize);
+  return kMpeg1BitrateTable[btr_idx][vidx*3+lidx] * 1000;
+}
+
+static inline size_t
+Mpeg1FrameSize(uint8_t layer, size_t bitrate, size_t sample_rate, uint8_t padded)
+{
+  RCHECK(sample_rate > 0);
+  if (layer == kMpeg1L_1)
+    return (12 * bitrate / sample_rate + padded) * 4;
+  return 144 * bitrate / sample_rate + padded;
+}
+
+}  // namespace
+
+namespace shaka {
+namespace media {
+namespace mp2t {
+
+bool Mpeg1Header::IsSyncWord(const uint8_t* buf) const {
+  return (buf[0] == 0xff)
+    && ((buf[1] & 0b11100000) == 0b11100000)
+    // Version 01 is reserved
+    && ((buf[1] & 0b00011000) != 0b00001000)
+    // Layer 00 is reserved
+    && ((buf[1] & 0b00000110) != 0b00000000);
+}
+
+size_t Mpeg1Header::GetMinFrameSize() const {
+  return kMpeg1HeaderMinSize + 1;
+}
+
+size_t Mpeg1Header::GetSamplesPerFrame() const {
+  DCHECK_GT(layer_, 0);
+  DCHECK_GT(3 - layer_, 0);
+  DCHECK_LT(3 - layer_, static_cast<int>(kMpeg1SamplesPerFrameTableSize));
+  return kMpeg1SamplesPerFrameTable[3 - layer_];
+}
+
+bool Mpeg1Header::Parse(const uint8_t* mpeg1_frame, size_t mpeg1_frame_size) {
+  DCHECK(mpeg1_frame);
+
+  if (mpeg1_frame_size < kMpeg1HeaderMinSize)
+    return false;
+
+  BitReader frame(mpeg1_frame, mpeg1_frame_size);
+  // Verify frame starts with sync bits (0x7ff).
+  uint32_t sync;
+  RCHECK(frame.ReadBits(11, &sync));
+  RCHECK(sync == 0x7ff);
+  // MPEG version and layer.
+  RCHECK(frame.ReadBits(2, &version_));
+  RCHECK(version_ != kMpeg1V_INV);
+  RCHECK(frame.ReadBits(2, &layer_));
+  RCHECK(layer_ != kMpeg1L_INV);
+  RCHECK(frame.ReadBits(1, &protection_absent_));
+
+  uint8_t btr_idx;
+  RCHECK(frame.ReadBits(4, &btr_idx));
+  RCHECK(btr_idx > 0);
+  bitrate_ = Mpeg1BitRate(btr_idx, version_, layer_);
+
+  uint8_t sr_idx;
+  RCHECK(frame.ReadBits(2, &sr_idx));
+  RCHECK(sr_idx < 0b11);
+  sample_rate_ = Mpeg1SampleRate(sr_idx, version_);
+
+  RCHECK(frame.ReadBits(1, &padded_));
+  // Skip private stream bit.
+  RCHECK(frame.SkipBits(1));
+
+  RCHECK(frame.ReadBits(2, &channel_mode_));
+  // Skip Mode extension
+  RCHECK(frame.SkipBits(2));
+  // Skip copyright, origination and emphasis info.
+  RCHECK(frame.SkipBits(4));
+
+  return true;
+}
+
+size_t Mpeg1Header::GetHeaderSize() const {
+  // Unlike ADTS, for MP3, the whole frame is included in the media sample, so
+  // return 0 header size.
+  return 0;
+}
+
+size_t Mpeg1Header::GetFrameSize() const {
+  return Mpeg1FrameSize(layer_, bitrate_, sample_rate_, padded_);
+}
+
+size_t Mpeg1Header::GetFrameSizeWithoutParsing(const uint8_t* data,
+                                               size_t num_bytes) const {
+  DCHECK_GT(num_bytes, static_cast<size_t>(2));
+  uint8_t version = (data[1] & 0b00011000) >> 3;
+  uint8_t layer = (data[1] & 0b00000110) >> 1;
+  uint8_t btr_idx = (data[2] & 0b11110000) >> 4;
+  uint8_t sr_idx = (data[2] & 0b00001100) >> 2;
+  uint8_t padded = (data[2] & 0b00000010) >> 1;
+
```

Comment:
Adding a check to avoid possible assertion in the computations below.

``` c++
if (version == kMpeg1V_INV || layer == kMpeg1L_INV || btr_idx == 0 || sr_idx == 0b11)
  return 0;
```